allo/backend/hls.py

```diff
@@ -471,6 +531,93 @@ def __call__(self, *args, shell=True):
             arr = np.fromfile(f"{self.project}/output.data", dtype=args[-1].dtype)
             args[-1][:] = arr.reshape(args[-1].shape)
             return
+        elif self.platform == "pynq":
+            # Do not assert PYNQ availability here; the presence of a physical
+            # PYNQ device should be checked by callers that need it.
+            if self.mode == "csim":
+                cwd = os.getcwd()
+                mod = IPModule(
+                    top=self.top_func_name,
+                    impl=f"{cwd}/{self.project}/kernel.cpp",
+                    link_hls=True,
+                )
+                mod(*args)
+                return
+            if self.mode == "csyn":
+                cmd = f"cd {self.project}; vitis_hls -f run.tcl"
+                assert len(args) == 0, "csyn mode does not need to pass in arguments"
+                print(
+                    f"[{time.strftime('%H:%M:%S', time.gmtime())}] Begin synthesizing project ..."
+                )
+                if shell:
+                    process = subprocess.Popen(cmd, shell=True)
+                else:
+                    process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
+                process.wait()
+                if process.returncode != 0:
+                    raise RuntimeError("Failed to synthesize the design")
+                # only run block design if configs['run_impl'] or mode contains
+                # 'impl'. Use safe checks instead of catching broad exceptions.
+                do_impl = False
+                if isinstance(self.configs, dict):
+                    do_impl = bool(self.configs.get("run_impl", False))
+                if not do_impl and not (
+                    isinstance(self.mode, str) and "impl" in self.mode
+                ):
+                    # return early if not running impl
+                    return
+
+                # vivado block design
+                bd_script = (
+                    self.custom_bd_tcl if self.custom_bd_tcl else "block_design.tcl"
+                )
```

Comment:
A bit messy here. Should set up default `self.custom_bd_tcl = "block_design.tcl"` when you first process this value.